bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceLogic.java

```diff
@@ -0,0 +1,799 @@
+				isWholeWordSearch());
+	}
+
+	/**
+	 * Replaces the selection with <code>replaceString</code>. If
+	 * <code>regExReplace</code> is <code>true</code>, <code>replaceString</code> is
+	 * a regex replace pattern which will get expanded if the underlying target
+	 * supports it. Returns the region of the inserted text; note that the returned
+	 * selection covers the expanded pattern in case of regex replace.
+	 *
+	 * @param replaceString the replace string (or a regex pattern)
+	 * @param regExReplace  <code>true</code> if <code>replaceString</code> is a
+	 *                      pattern
+	 * @return the selection after replacing, i.e. the inserted text
+	 * @since 3.0
+	 */
+	public Point replaceSelection(String replaceString, boolean regExReplace) {
+		if (target instanceof IFindReplaceTargetExtension3)
+			((IFindReplaceTargetExtension3) target).replaceSelection(replaceString, regExReplace);
+		else
+			target.replaceSelection(replaceString);
+
+		return target.getSelection();
+	}
+
+	/**
+	 * Returns whether the specified search string can be found using the given
+	 * options.
+	 *
+	 * @param findString    the string to search for
+	 * @param forwardSearch the direction of the search
+	 * @return <code>true</code> if the search string can be found using the given
+	 *         options
+	 *
+	 * @since 3.0
+	 */
+	public boolean findNext(String findString, boolean forwardSearch) {
+
+		if (target == null)
+			return false;
+
+		Point r = null;
+		if (isIncrementalSearch())
+			r = incrementalBaseLocation;
+		else
+			r = target.getSelection();
+
+		int findReplacePosition = r.x;
+		if (forwardSearch && !nextReplactionOperationNeedsFindOperationFirst || !forwardSearch && nextReplactionOperationNeedsFindOperationFirst)
+			findReplacePosition += r.y;
+
+		nextReplactionOperationNeedsFindOperationFirst = false;
+
+		int index = findIndex(findString, findReplacePosition);
+
+		if (index == -1) {
+			String msg = NLSUtility.format(EditorMessages.FindReplace_Status_noMatchWithValue_label, findString);
+			statusMessage(false, EditorMessages.FindReplace_Status_noMatch_label, msg);
+			return false;
+		}
+
+		if (forwardSearch && index >= findReplacePosition || !forwardSearch && index <= findReplacePosition)
+			statusMessage(""); //$NON-NLS-1$
+
+		return true;
+	}
+
+	public boolean performFindFirstThenReplaceInASecondStep(String findString, String replaceString) {
+		if (nextReplactionOperationNeedsFindOperationFirst) {
+			performSearch(findString);
+		}
+		if (performReplaceSelection(replaceString)) {
+			performSearch(findString);
+			return true;
+		}
+		return false;
+	}
+
+	public boolean performSelectAndReplace(String findString, String replaceString) {
+		if (nextReplactionOperationNeedsFindOperationFirst)
+			performSearch(findString);
+		return performReplaceSelection(replaceString);
+	}
+
+	public void updateTarget(IFindReplaceTarget newRarget, boolean canEditTarget) {
```

Comment:
Would be nice to correct this typo `newRarget` -> `newTarget` in this refactoring.